ui/src/app/shared/components/parameters-input/parameters-input.tsx

```diff
@@ -42,39 +21,55 @@ export class ParametersInput extends React.Component<ParametersInputProps, {para
                     title: value
                 }))}
                 onChange={e => {
-                    const newParameters: Parameter[] = this.state.parameters.map(p => ({
+                    const newParameters: Parameter[] = parameters.map(p => ({
                         name: p.name,
                         value: p.name === parameter.name ? e.value : Utils.getValueFromParameter(p),
                         enum: p.enum
                     }));
-                    this.setState({parameters: newParameters});
-                    this.onParametersChange(newParameters);
+                    setParameters(newParameters);
+                    onParametersChange(newParameters);
                 }}
             />
         );
     }
 
-    private displayInputFieldForSingleValue(parameter: Parameter) {
+    function displayInputFieldForSingleValue(parameter: Parameter) {
         return (
             <textarea
                 className='argo-field'
                 value={Utils.getValueFromParameter(parameter)}
                 onChange={e => {
-                    const newParameters: Parameter[] = this.state.parameters.map(p => ({
+                    const newParameters: Parameter[] = parameters.map(p => ({
                         name: p.name,
                         value: p.name === parameter.name ? e.target.value : Utils.getValueFromParameter(p),
                         enum: p.enum
                     }));
-                    this.setState({parameters: newParameters});
-                    this.onParametersChange(newParameters);
+                    setParameters(newParameters);
+                    onParametersChange(newParameters);
                 }}
             />
         );
     }
 
-    private onParametersChange(parameters: Parameter[]) {
-        if (this.props.onChange) {
-            this.props.onChange(parameters);
+    function onParametersChange(params: Parameter[]) {
+        if (props.onChange) {
+            props.onChange(params);
         }
     }
```

Comment:
```suggestion
    function onParameterChange(parameter: Parameter, value: string) {
        const newParameters: Parameter[] = props.parameters.map(p => ({
            name: p.name,
            value: p.name === parameter.name ? value : Utils.getValueFromParameter(p),
            enum: p.enum
        }));
        props.onChange(newParameters);
    }
```

we can consolidate the change logic into this function. I also made it singular as this is a function that runs when a single parameter changes. (the previous name confused me a bit until I rewrote it)

Reply:
Could you also move this function to the top of the file? Then it would have consistent ordering with other components in the codebase where they have `state`, then callbacks, then render functions

Reply:
> we can consolidate the change logic into this function. I also changed the name to be singular as this is a function that runs when a _single_ parameter changes. (the previous name confused me a bit until I rewrote it)

I realize that it is important to change with good functions name!

Reply:
> Ah! It means that `onParamterChange` functions is callbacks, and other display functions is render function right?

Yep! They're pretty common patterns in React components